File: dotnet/Lexy/Lexy.Compiler/Language/Node.cs
using System;
using System.Collections.Generic;
using Lexy.Compiler.Parser;

namespace Lexy.Compiler.Language
{
    public abstract class Node : INode
    {
        public SourceReference Reference { get; }

        protected Node(SourceReference reference)
        {
            Reference = reference ?? throw new ArgumentNullException(nameof(reference));
        }

        public virtual void ValidateTree(IValidationContext context)
        {
            Validate(context);

            var children = GetChildren();
            foreach (var child in children)
            {
                ValidateNodeTree(context, child);
            }
        }

        protected virtual void ValidateNodeTree(IValidationContext context, INode child)
        {
            if (child == null) throw new InvalidOperationException($"({GetType().Name}) Child is null");

            if (child is IRootNode node)
            {
                context.Logger.SetCurrentNode(node);
            }

            child.ValidateTree(context);

            if (this is IRootNode)
            {
                context.Logger.SetCurrentNode(this as IRootNode);
            }
        }

        public abstract IEnumerable<INode> GetChildren();

        protected abstract void Validate(IValidationContext context);
    }
}

File: typescipt/src/language/node.ts
import {IValidationContext} from "../parser/validationContext";
import {SourceReference} from "../parser/sourceReference";
import {IRootNode} from "./RootNode";
import {nameOf} from "../infrastructure/nameOf";
import {ITypeWithMembers} from "./types/iTypeWithMembers";

export interface INode {
  reference: SourceReference

  validateTree(context: IValidationContext): void

  getChildren(): Array<INode>;
}

export abstract class Node implements INode {
  
  public reference: SourceReference;

  protected constructor(reference: SourceReference) {
     this.reference = reference ;
   }

   public validateTree(context: IValidationContext): void {
     this.validate(context);

     let children = this.getChildren();
     children.forEach(child => this.validateNodeTree(context, child));
   }

  public abstract nodeType: string;
  public abstract getChildren(): Array<INode>;

   protected validateNodeTree(context: IValidationContext, child: INode | null): void {
     if (child == null) throw new Error(`(${this.nodeType}) Child is null`);

     const rootNode = nameOf<IRootNode>("isRootNode") in this ? child as IRootNode : null;
     if (rootNode != null) {
       context.logger.setCurrentNode(rootNode);
     }

     child.validateTree(context);

     const thisAsRootNode = nameOf<IRootNode>("isRootNode") in this ? this as IRootNode : null;
     if (thisAsRootNode != null) {
       context.logger.setCurrentNode(thisAsRootNode);
     }
   }

   protected abstract validate(context: IValidationContext): void;
}
